services/users/session.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package users
+
+import (
+	"sync"
+	"time"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+var userSessionPool = sync.Pool{
+	New: func() interface{} {
+		return &model.Session{}
+	},
+}
+
+func ReturnSessionToPool(session *model.Session) {
+	if session != nil {
+		session.Id = ""
+		userSessionPool.Put(session)
+	}
+}
+
+func (us *UserService) CreateSession(session *model.Session) (*model.Session, error) {
+	session.Token = ""
+
+	session, err := us.sessionStore.Save(session)
+	if err != nil {
+		return nil, err
+	}
+
+	us.AddSessionToCache(session)
+
+	return session, nil
+}
+
+func (us *UserService) GetSession(token string) (*model.Session, error) {
+	var session = userSessionPool.Get().(*model.Session)
+	if err := us.sessionCache.Get(token, session); err == nil {
+		if us.metrics != nil {
+			us.metrics.IncrementMemCacheHitCounterSession()
+		}
+	} else {
+		if us.metrics != nil {
+			us.metrics.IncrementMemCacheMissCounterSession()
+		}
+	}
+	return session, nil
+}
+
+func (us *UserService) AddSessionToCache(session *model.Session) {
+	us.sessionCache.SetWithExpiry(session.Token, session, time.Duration(int64(*us.config().ServiceSettings.SessionCacheInMinutes))*time.Minute)
+}
+
+func (us *UserService) SessionCacheLength() int {
+	if l, err := us.sessionCache.Len(); err == nil {
+		return l
+	}
+	return 0
+}
+
+func (us *UserService) ClearSessionCacheForUserSkipClusterSend(userID string) {
```

Comment:
I'd like to change the naming from "skipclustersend" to "local". Thinking of something like this:
```
ClearUserSessionCache
ClearUserSessionCacheLocal
ClearAllUsersSessionCache
ClearAllUsersSessionCacheLocal
```